smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,290 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smack.websocket;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.net.SocketFactory;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+
+import org.jivesoftware.smack.SmackException;
+import org.jivesoftware.smack.SmackFuture;
+import org.jivesoftware.smack.SmackFuture.InternalSmackFuture;
+import org.jivesoftware.smack.XMPPException;
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnection.ConnectedButUnauthenticatedStateDescriptor;
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnection.LookupRemoteConnectionEndpointsStateDescriptor;
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnectionModule;
+import org.jivesoftware.smack.c2s.XmppClientToServerTransport;
+import org.jivesoftware.smack.c2s.internal.ModularXmppClientToServerConnectionInternal;
+import org.jivesoftware.smack.c2s.internal.WalkStateGraphContext;
+import org.jivesoftware.smack.datatypes.UInt16;
+import org.jivesoftware.smack.fsm.State;
+import org.jivesoftware.smack.fsm.StateDescriptor;
+import org.jivesoftware.smack.fsm.StateTransitionResult;
+import org.jivesoftware.smack.fsm.StateTransitionResult.AttemptResult;
+import org.jivesoftware.smack.packet.TopLevelStreamElement;
+import org.jivesoftware.smack.util.rce.RemoteConnectionEndpointLookupFailure;
+import org.jivesoftware.smack.websocket.XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints;
+import org.jivesoftware.smack.websocket.elements.OpenElement;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpoint;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpointFinder;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpointFinder.Result;
+
+public final class XmppWebsocketTransportModule
+                extends ModularXmppClientToServerConnectionModule<XmppWebsocketTransportModuleDescriptor> {
+    private final XmppWebsocketTransport websocketTransport;
+
+    private WebsocketConnectionAttemptState connectionAttemptState;
+
+    protected XmppWebsocketTransportModule(XmppWebsocketTransportModuleDescriptor moduleDescriptor,
+                    ModularXmppClientToServerConnectionInternal connectionInternal) {
+        super(moduleDescriptor, connectionInternal);
+
+        websocketTransport = new XmppWebsocketTransport(connectionInternal);
+    }
+
+    private void afterOutgoingElementsQueueModified() {
+        // After outgoing queue is modified, send the top level stream element obtained by polling.
```

Comment:
Inline this method where it belongs. It is also not entirely correct, but more on that later.